### ISSUE TITLE
json.exception.parse_error on nixos-unstable-small

### ISSUE THREAD

**Issue by lovesegfault** [nix-community/nix-direnv#113]
Using `use flake` on my [config repo](https://github.com/lovesegfault/nix-config)'s `flake-update` branch causes the following error to appear:
```console
$ direnv reload
direnv: loading ~/src/nix-config/.envrc                                                                                                                                                               
direnv: using flake
warning: Git tree '/home/bemeurer/src/nix-config' is dirty
error: [json.exception.parse_error.101] parse error at line 57, column 178: syntax error while parsing value - invalid string: ill-formed UTF-8 byte; last read: '" \n    local fn=\"$1\";\n    local fd;\n    local magic;\n    exec {fd}< \"$fn\";\n    read -r -n 4 -u \"$fd\" magic;\n    exec {fd}>&-;\n    if [[ \"$magic\" = ''
warning: Git tree '/home/bemeurer/src/nix-config' is dirty
direnv: renewed cache
direnv: export ~XDG_DATA_DIRS
```

The first thing that strikes me is that it complains about my tree being dirty, even when it isn't:
```console
$ git status   
On branch flake-update
Your branch is up to date with 'origin/flake-update'.

nothing to commit, working tree clean
```

If I replace `use flake` with `use nix` things seem to mostly work fine, so I suspect a bump of `nixUnstable` might have broken things.

This is running off of commit `a878fd020e1ab88888ef72aae46d2c98d9212c79` on my repo, which is using nixpkgs commit `ff748a82a5fff004c952f32eacd64604a6f4036f`.

**Comment by lovesegfault** [comment]
The issue happens somewhere between the following nixpkgs commits:

* Good: `0171610faad12bdc6757b483b7e7219780123ff1`
* Bad: `ff748a82a5fff004c952f32eacd64604a6f4036f`

**Comment by lovesegfault** [comment]
* Good: `0171610faad12bdc6757b483b7e7219780123ff1`
* Bad: `ff748a82a5fff004c952f32eacd64604a6f4036f`

**Comment by winterqt** [comment]
> * Good: `0171610faad12bdc6757b483b7e7219780123ff1`

That's... strange, to say the least, as https://github.com/NixOS/nixpkgs/commit/04b552c32312e8067ab5e3948513ce4624f35c4d was the last time `nixUnstable` was bumped, and that was before this commit. I wonder why that is...

**Comment by winterqt** [comment]
This is weirder than I thought.

I tested using commit `2096a1678f6b1bba91e01410b5baac0fc2cde144`, which is right before the last `nixUnstable` bump. I only got `use flake` to work if I pinned my project flake to that commit, what my system flake is on (`2bb5cbf7f8b99a8d1d6646abe5ab993f6823212f`, the latest commit on `nixpkgs-unstable` as of writing)

So I don't *think* this is a Nix issue?

**Comment by winterqt** [comment]
> `error: [json.exception.parse_error.101] parse error at line 57, column 178: syntax error while parsing value - invalid string: ill-formed UTF-8 byte; last read: '" \n    local fn=\"$1\";\n    local fd;\n    local magic;\n    exec {fd}< \"$fn\";\n    read -r -n 4 -u \"$fd\" magic;\n    exec {fd}>&-;\n    if [[ \"$magic\" = ''`

This is [this line](https://github.com/NixOS/nixpkgs/blob/78230d1c42c8f0e870db6b9a8856c10c2826ee94/pkgs/stdenv/generic/setup.sh#L211) in the stdenv setup script, for what it's worth.

**Comment by Mic92** [comment]
So this only happen on macOS?

**Comment by winterqt** [comment]
I'm not sure, since it's not even executing the function, it's just failing to read the derivation JSON(?)

**Comment by lovesegfault** [comment]
> So this only happen on macOS?
Definitely not, I'm on NixOS

**Comment by winterqt** [comment]
Here's what I've been able to figure out so far:

- The command that's failing is `print-dev-env`
- It's failing when reading the `nix-shell-env` file in the store, because there's invalid UTF-8 bytes at the line.

<img width="841" alt="image" src="https://user-images.githubusercontent.com/78392041/133645548-c00e132a-c087-4b79-ab41-4388dd8dcbc9.png">

```bash
elif [[ "$magic" = $'\xca\xfe\xba\xbe' || "$magic" = $'\xbe\xba\xfe\xca' ]]; then
```

Instead of `\xca\fe\ba\be` being added as a UTF-8 string, the literal bytes `CA FE BA BE` are being inserted into the file at that point, during the build process.

**Comment by lovesegfault** [comment]
> Here's what I've been able to figure out so far:
> 
>     * The command that's failing is `print-dev-env`
> 
>     * It's failing when reading the `nix-shell-env` file in the store, because there's invalid UTF-8 bytes at the line in question:
> 
> 
> <img alt="image" width="841" src="https://user-images.githubusercontent.com/78392041/133645548-c00e132a-c087-4b79-ab41-4388dd8dcbc9.png">
> 
> ```shell
> elif [[ "$magic" = $'\xca\xfe\xba\xbe' || "$magic" = $'\xbe\xba\xfe\xca' ]]; then
> ```
> 
> So, instead of `\xca\fe\ba\be` being added as a UTF-8 string, the literal bytes `CA FE BA BE` are being inserted into the file at that location during the build process.

I'm still bisecting the commit range I mentioned above to see if we can pin down a guilty commit, but it involves building GHC a billion times

**Comment by winterqt** [comment]
> but it involves building GHC a billion times

My condolences to your machine. 😅 that doesn't sound fun.

This seems to be an issue in Nix's `BuildEnvironment` class, and I'm also currently trying to see where this bug lies exactly, and where it was introduced.

**Comment by winterqt** [comment]
The bad commit is https://github.com/NixOS/nixpkgs/commit/488395c0f80e2863e9cf83ffb029a330c8e19573.

**Comment by dschrempf** [comment]
For the record: If I create a new directory with a direnv setup, I get the JSON parse error mentioned above. However, on other, already established projects, I get some weird evalution error:
```
direnv: loading ~/Shared/haskell/elynx/.envrc
direnv: using flake
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint:
hint: 	git config --global init.defaultBranch <name>
hint:
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint:
hint: 	git branch -m <name>
/nix/store/5nx9b7qn333asgay49vgrb01n5jjdkyk-nix-direnv-1.4.0/share/nix-direnv/direnvrc: eval:1793: syntax error: unexpected end of file
direnv: export +AR +AR_FOR_TARGET +AS +AS_FOR_TARGET +CC +CC_FOR_TARGET +CONFIG_SHELL +CXX +CXX_FOR_TARGET +HOST_PATH +IN_NIX_SHELL +LD +LD_FOR_TARGET +NIX_BINTOOLS +NIX_BINTOOLS_FOR_TARGET +NIX_BINTOOLS_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_BINTOOLS_WRAPPER_TARGET_TARGET_x86_64_unknown_linux_gnu +NIX_BUILD_CORES +NIX_CC +NIX_CC_FOR_TARGET +NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_CC_WRAPPER_TARGET_TARGET_x86_64_unknown_linux_gnu +NIX_CFLAGS_COMPILE +NIX_CFLAGS_COMPILE_FOR_TARGET +NIX_ENFORCE_NO_NATIVE +NIX_GHCPKG +NIX_GHC_DOCDIR +NIX_GHC_LIBDIR +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_LDFLAGS_FOR_TARGET +NIX_STORE +NM +NM_FOR_TARGET +OBJCOPY +OBJCOPY_FOR_TARGET +OBJDUMP +OBJDUMP_FOR_TARGET +RANLIB +RANLIB_FOR_TARGET +READELF +READELF_FOR_TARGET +SIZE +SIZE_FOR_TARGET +SOURCE_DATE_EPOCH +STRINGS +STRINGS_FOR_TARGET +STRIP +STRIP_FOR_TARGET +buildInputs +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +dontAddDisableDepTrack +installPhase +name +nativeBuildInputs +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~LOCALE_ARCHIVE ~NIX_GHC ~PATH ~XDG_DATA_DIRS
```
Weird that git complains about the default branch not being set.

**Comment by winterqt** [comment]
I believe I have found a fix for this issue, I'm just waiting on Nix to compile.

**Comment by lovesegfault** [comment]
Weird, `iconv` reports that file as being UTF-8:
```console
$ iconv -f UTF-8 pkgs/stdenv/generic/setup.sh -t UTF-8 -o /dev/null
$ echo $?                                                          
0
```

**Comment by winterqt** [comment]
After the last few hours debugging, it turns out that my initial assumption of where the issue lies is wrong.

Basically: when Nix is building the shell environment derivation, it runs¹ [this script](https://github.com/NixOS/nix/blob/1ec4efa6c82fb6d90887c21aa7c7c6f3c644dd65/src/nix/get-env.sh) to generate the JSON that's failing to be parsed. There's a function there called `__escapeString`, which escapes things like `"` and `\`. So, I tried to patch that script to include `\x`. However:
- this case is already covered by `__s="${__s//\\/\\\\}"`
- `get-env.sh` only has the malformed setup script to work on (where the hex literals are replaced by their actual bytes (`\xCA` -> `0xCA`)

So the problem is somewhere before it, and I really don't know where that could be.

¹ Well technically the derivation build process does, but I'm simplifying things. ~~(I don't really fully understand it myself)~~

**Comment by winterqt** [comment]
Okay, I figured it out.

Our use of `declare -F` to find the current loaded functions in `get-env` makes bash interpret the hex literals.

```bash
$ x(){
echo $'\xAB'
}
$ $(declare -F)
x ()
{
    echo '�'
}
```

**Comment by winterqt** [comment]
I've also opened https://github.com/NixOS/nixpkgs/pull/138186, so at least `nix-direnv` will work while the Nix issue is solved.

**Comment by winterqt** [comment]
I've made this patch as a stopgap, so we can use `nix-direnv` while the various fixes are being sorted out:

```diff
diff --git a/src/nix/get-env.sh b/src/nix/get-env.sh
index 42c806450..a8563c772 100644
--- a/src/nix/get-env.sh
+++ b/src/nix/get-env.sh
@@ -8,6 +8,8 @@ if [[ -n $stdenv ]]; then
     source $stdenv/setup
 fi
 
+unset -f isMachO
+
 # Better to use compgen, but stdenv bash doesn't have it.
 __vars="$(declare -p)"
 __functions="$(declare -F)"
```

You can apply it to `nixUnstable` using an overlay:

```nix
(self: super:
{
  nixUnstable = super.nixUnstable.override {
    patches = [ ./unset-is-macho.patch ];
  };
})
```

**Comment by dschrempf** [comment]
Thanks for your prompt response and excellent work on this!

**Comment by jrolfs** [comment]
For anyone else using this patch with **nix-darwin**, I had to disable the **zsh** completions ([`zsh.enableCompletion`](https://github.com/LnL7/nix-darwin/blob/a7492a8c76dcc702d0a65cd820a5f9baa8702684/modules/programs/zsh/default.nix#L61)) like [**nixpkgs**](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/zsh/zsh.nix#L289-L296) does.

**Comment by dhess** [comment]
> For anyone else using this patch with **nix-darwin**, I had to disable the **zsh** completions ([`zsh.enableCompletion`](https://github.com/LnL7/nix-darwin/blob/a7492a8c76dcc702d0a65cd820a5f9baa8702684/modules/programs/zsh/default.nix#L61)) like [**nixpkgs**](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/zsh/zsh.nix#L289-L296) does.

I had that issue and gave up in frustration and confusion. Do you have any idea why on earth it's necessary?

**Comment by dschrempf** [comment]
May I ask about the status of this bug? I see that commit https://github.com/NixOS/nixpkgs/pull/138186/commits/f8edf7720d7ddcc1e9c85f9a00f1797b1112fbdf is now in `nixos-unstable`, but direnv still fails without applying your patch to `nixUnstable`. Thanks!

**Comment by toastal** [comment]
@dschrempf This prompted me to try it, but

```console
$ rm -rf .direnv/
```

Seemed to allow me to finally `nix flake update` my project with success -- albeit needing to redownload all dependencies.

**Comment by dschrempf** [comment]
Hm, on `nixos-unstable` (updated just now), and without the patch, I get, for example,
```
direnv: loading ~/Shared/haskell/circular/.envrc
direnv: using flake
direnv: renewed cache
/nix/store/yna90284zsc2lykjf1v19idcj3vhh5nb-nix-direnv-1.4.0/share/nix-direnv/direnvrc: eval:1793: syntax error: unexpected end of file
```

With the patch, everything works fine.

**Comment by winterqt** [comment]
@dschrempf Can you try running:
```bash
$ direnv deny
$ rm -r .direnv # or remove this directory's cache folder if you set a custom one
$ nix-collect-garbage -d
```
And then see if `nix develop` works?

**Comment by jrolfs** [comment]
@dhess my assumption was that the latest version of **nix** provides its own completions and therefore **nix-zsh-completions** conflicts with those. That's just an assumption though, I'm still largely a Nix noob 😅.

**Comment by dhess** [comment]
> @dhess my assumption was that the latest version of **nix** provides its own completions and therefore **nix-zsh-completions** conflicts with those. That's just an assumption though, I'm still largely a Nix noob 😅.

In my case, I was patching the current version of `nixUnstable`. It installs fine with no `zsh` completion conflicts... until I add the patch, and then it starts conflicting. The thing that doesn't make sense is that the patch has nothing to do with `zsh` completion :)

**Comment by dschrempf** [comment]
@winterqt I am still getting this error.
```
direnv: loading ~/Shared/haskell/circular/.envrc
direnv: using flake
[1/5 built, 1/6/8 copied (14.1 MiB), 2.2 MiB DL] fetching quickcheck-instances-0.3.25.2-doc from https://cache.nixos.orgdirenv: ([/nix/store/qjwf74m3wi2glya8qig88zn8cpr18w5f-direnv-2.28.0/bin/direnv export bash]) is taking a while to execute. Use CTRL-C to give up.
direnv: renewed cache
/nix/store/yna90284zsc2lykjf1v19idcj3vhh5nb-nix-direnv-1.4.0/share/nix-direnv/direnvrc: eval:1793: syntax error: unexpected end of file
direnv: export +AR +AR_FOR_TARGET +AS +AS_FOR_TARGET +CC +CC_FOR_TARGET +CONFIG_SHELL +CXX +CXX_FOR_TARGET +HOST_PATH +IN_NIX_SHELL +LD +LD_FOR_TARGET +NIX_BINTOOLS +NIX_BINTOOLS_FOR_TARGET +NIX_BINTOOLS_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_BINTOOLS_WRAPPER_TARGET_TARGET_x86_64_unknown_linux_gnu +NIX_BUILD_CORES +NIX_CC +NIX_CC_FOR_TARGET +NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_CC_WRAPPER_TARGET_TARGET_x86_64_unknown_linux_gnu +NIX_CFLAGS_COMPILE +NIX_CFLAGS_COMPILE_FOR_TARGET +NIX_ENFORCE_NO_NATIVE +NIX_GHC +NIX_GHCPKG +NIX_GHC_DOCDIR +NIX_GHC_LIBDIR +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_LDFLAGS_FOR_TARGET +NIX_STORE +NM +NM_FOR_TARGET +OBJCOPY +OBJCOPY_FOR_TARGET +OBJDUMP +OBJDUMP_FOR_TARGET +RANLIB +RANLIB_FOR_TARGET +READELF +READELF_FOR_TARGET +SIZE +SIZE_FOR_TARGET +SOURCE_DATE_EPOCH +STRINGS +STRINGS_FOR_TARGET +STRIP +STRIP_FOR_TARGET +buildInputs +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +dontAddDisableDepTrack +installPhase +name +nativeBuildInputs +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~LOCALE_ARCHIVE ~PATH ~XDG_DATA_DIRS
The program 'jump' is not in your PATH. You can make it available in an
ephemeral shell by typing:
  nix-shell -p jump
```
First it correctly builds the environment, but then when it activates it, I get the syntax error.

I performed garbage collection after deleting the `.direnv` file, I even rebooted my computer before entering the directory. No idea what's going on here.

**Comment by winterqt** [comment]
> @winterqt I am still getting this error.

@dschrempf Interesting.

Do you mind trying to see if `nix develop` works after GCing? I want to see if this is a `nix-direnv` issue or not. Also, do you mind sharing your `flake.nix`, or at least the parts with your `devShell` and what `nixpkgs-unstable` commit you're on?

This doesn't seem like the same issue since `nixpkgs-unstable` commit doesn't include https://github.com/NixOS/nixpkgs/pull/138334 yet, so we may want to split off into a separate issue.

**Comment by dschrempf** [comment]
OK, so it is already the `nix develop` command that fails.
```
 + nix develop
bash: /tmp/nix-shell.MhRjQ5: line 1647: syntax error: unexpected end of file
```
I am not sure if this shell command contains private data, so I am not uploading it but let me know it should be provided. The flake looks like so:
```nix
{
  description = "Circular stacks";

  inputs.flake-utils.url = "github:numtide/flake-utils";

  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";

  outputs =
    { self
    , flake-utils
    , nixpkgs
    }:
      flake-utils.lib.eachDefaultSystem (
        system:
          let
            pkgs = import nixpkgs { inherit system; };
            haskellPackages = pkgs.haskellPackages;
            circular = haskellPackages.callCabal2nix "circular" self rec {};
            circular-dev = pkgs.haskell.lib.doBenchmark circular;
          in
            {
              packages.circular = circular;

              defaultPackage = circular;

              devShell = pkgs.haskellPackages.shellFor {
                packages = _: [ circular-dev ];
                buildInputs = with pkgs; [
                  bashInteractive
                  haskellPackages.cabal-install
                  haskellPackages.haskell-language-server
                  haskellPackages.stack
                ];
                doBenchmark = true;
                withHoogle = true;
              };
            }
      );
}
```

BTW: I removed all my .direnv folders now, and performed a subsequent garbage collection step. But maybe, there are still some old builds lingering around? I don't know.

**Comment by theduke** [comment]
Is this supposed to work on unstable now?

Because I just tried a rebuild without the patch, and it's still not working for me.

**Comment by Mic92** [comment]
It's still in staging-next.

**Comment by Mic92** [comment]
> OK, so it is already the `nix develop` command that fails.
> 
> ```
>  + nix develop
> bash: /tmp/nix-shell.MhRjQ5: line 1647: syntax error: unexpected end of file
> ```
> 
> I am not sure if this shell command contains private data, so I am not uploading it but let me know it should be provided. The flake looks like so:
> 
> ```nix
> {
>   description = "Circular stacks";
> 
>   inputs.flake-utils.url = "github:numtide/flake-utils";
> 
>   inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
> 
>   outputs =
>     { self
>     , flake-utils
>     , nixpkgs
>     }:
>       flake-utils.lib.eachDefaultSystem (
>         system:
>           let
>             pkgs = import nixpkgs { inherit system; };
>             haskellPackages = pkgs.haskellPackages;
>             circular = haskellPackages.callCabal2nix "circular" self rec {};
>             circular-dev = pkgs.haskell.lib.doBenchmark circular;
>           in
>             {
>               packages.circular = circular;
> 
>               defaultPackage = circular;
> 
>               devShell = pkgs.haskellPackages.shellFor {
>                 packages = _: [ circular-dev ];
>                 buildInputs = with pkgs; [
>                   bashInteractive
>                   haskellPackages.cabal-install
>                   haskellPackages.haskell-language-server
>                   haskellPackages.stack
>                 ];
>                 doBenchmark = true;
>                 withHoogle = true;
>               };
>             }
>       );
> }
> ```
> 
> BTW: I removed all my .direnv folders now, and performed a subsequent garbage collection step. But maybe, there are still some old builds lingering around? I don't know.
> 
> EDIT: You are right, this is maybe a different issue, let me know if I should open one. However, it can be solved using the same patch.



> @winterqt I am still getting this error.
> 
> ```
> direnv: loading ~/Shared/haskell/circular/.envrc
> direnv: using flake
> [1/5 built, 1/6/8 copied (14.1 MiB), 2.2 MiB DL] fetching quickcheck-instances-0.3.25.2-doc from https://cache.nixos.orgdirenv: ([/nix/store/qjwf74m3wi2glya8qig88zn8cpr18w5f-direnv-2.28.0/bin/direnv export bash]) is taking a while to execute. Use CTRL-C to give up.
> direnv: renewed cache
> /nix/store/yna90284zsc2lykjf1v19idcj3vhh5nb-nix-direnv-1.4.0/share/nix-direnv/direnvrc: eval:1793: syntax error: unexpected end of file
> direnv: export +AR +AR_FOR_TARGET +AS +AS_FOR_TARGET +CC +CC_FOR_TARGET +CONFIG_SHELL +CXX +CXX_FOR_TARGET +HOST_PATH +IN_NIX_SHELL +LD +LD_FOR_TARGET +NIX_BINTOOLS +NIX_BINTOOLS_FOR_TARGET +NIX_BINTOOLS_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_BINTOOLS_WRAPPER_TARGET_TARGET_x86_64_unknown_linux_gnu +NIX_BUILD_CORES +NIX_CC +NIX_CC_FOR_TARGET +NIX_CC_WRAPPER_TARGET_HOST_x86_64_unknown_linux_gnu +NIX_CC_WRAPPER_TARGET_TARGET_x86_64_unknown_linux_gnu +NIX_CFLAGS_COMPILE +NIX_CFLAGS_COMPILE_FOR_TARGET +NIX_ENFORCE_NO_NATIVE +NIX_GHC +NIX_GHCPKG +NIX_GHC_DOCDIR +NIX_GHC_LIBDIR +NIX_HARDENING_ENABLE +NIX_INDENT_MAKE +NIX_LDFLAGS +NIX_LDFLAGS_FOR_TARGET +NIX_STORE +NM +NM_FOR_TARGET +OBJCOPY +OBJCOPY_FOR_TARGET +OBJDUMP +OBJDUMP_FOR_TARGET +RANLIB +RANLIB_FOR_TARGET +READELF +READELF_FOR_TARGET +SIZE +SIZE_FOR_TARGET +SOURCE_DATE_EPOCH +STRINGS +STRINGS_FOR_TARGET +STRIP +STRIP_FOR_TARGET +buildInputs +builder +configureFlags +depsBuildBuild +depsBuildBuildPropagated +depsBuildTarget +depsBuildTargetPropagated +depsHostHost +depsHostHostPropagated +depsTargetTarget +depsTargetTargetPropagated +doCheck +doInstallCheck +dontAddDisableDepTrack +installPhase +name +nativeBuildInputs +out +outputs +patches +phases +propagatedBuildInputs +propagatedNativeBuildInputs +shell +shellHook +stdenv +strictDeps +system ~LOCALE_ARCHIVE ~PATH ~XDG_DATA_DIRS
> The program 'jump' is not in your PATH. You can make it available in an
> ephemeral shell by typing:
>   nix-shell -p jump
> ```
> 
> First it correctly builds the environment, but then when it activates it, I get the syntax error.
> 
> I performed garbage collection after deleting the `.direnv` file, I even rebooted my computer before entering the directory. No idea what's going on here.
> 
> EDIT: Maybe I should say that I use Home Manager to activate direnv.
> 
> ```
>   programs.direnv.enable = true;
>   programs.direnv.nix-direnv.enable = true;
>   programs.direnv.nix-direnv.enableFlakes = true;
> ```

You might still use the unpatched version than.

**Comment by winterqt** [comment]
> You might still use the unpatched version than.

The reason I say that this may be a different issue is that the error that they're getting seemingly has nothing to do with the error in this issue.
However, they said that the `unset isMachO` patch fixes it, so I'm a bit confused.

**Comment by dschrempf** [comment]
Oh OK, I didn't know it was only in `staging-next`. I was looking for the commit disabling the `machO` function, and I saw it in `nixos-unstable`. Maybe I confused something...

**Comment by dhess** [comment]
I'm running the latest `nixpkgs-unstable`, which incorporates the alleged `isMachO` fix, and I'm still seeing this error in most of my Flakes, as well.

Any idea what's going on?

**Comment by winterqt** [comment]
@dhess Strange. What error are you receiving specifically, the one in this issues body? It *should* at least be a different error, since the hex literals that were causing the error are no longer in stdenv.

**Comment by winterqt** [comment]
That commit [does not incorporate](https://github.com/NixOS/nixpkgs/blob/a3a23d9599b0a82e333ad91db2cdc479313ce154/pkgs/stdenv/generic/setup.sh) the fix.

**Comment by dhess** [comment]
Hmm, that's strange. `8cff7796d796c2c1e687657346db4abe5677cd09` comes after it in the `git log`, so I assumed it had been resolved. Sorry for the noise.

**Comment by dhess** [comment]
Now I'm up-to-date on `nixpkgs-unstable` (b5182c214fac1e6db9f28ed8a7cfc2d0c255c763). This commit includes the fix, right?

https://github.com/NixOS/nixpkgs/blame/b5182c214fac1e6db9f28ed8a7cfc2d0c255c763/pkgs/stdenv/generic/setup.sh#L201 include

I'm still having problems:

```
direnv: loading ~/git/hackworth-ops/.envrc
direnv: using flake
error: [json.exception.parse_error.101] parse error at line 46, column 178: syntax error while parsing value - invalid string: ill-formed UTF-8 byte; last read: '" \n    local fn=\"$1\";\n    local fd;\n    local magic;\n    exec {fd}< \"$fn\";\n    read -r -n 4 -u \"$fd\" magic;\n    exec {fd}>&-;\n    if [[ \"$magic\" = '?'
direnv: renewed cache
direnv: export ~XDG_DATA_DIRS
```

**Comment by winterqt** [comment]
@dhess Try clearing direnv cache + GCing and see if that fixes it. (*crosses fingers*)

**Comment by dhess** [comment]
Unfortunately, neither suggestions fixed the issue. It's not a big deal, I can use the `nix-shell` compatibility via `flake-compat` until this is sorted out.

**Comment by winterqt** [comment]
@lovesegfault I think we should be fine closing this issue by this point.
